packages/langium/src/documents/document-builder.ts

```diff
@@ -54,55 +50,72 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
+            if (this.connection && !cancelToken.isCancellationRequested) {
+                // Send the computed diagnostics to VS Code.
+                this.connection.sendDiagnostics({ uri: document.textDocument.uri, diagnostics });
+            }
+        }
+        if (!cancelToken.isCancellationRequested) {
+            document.state = DocumentState.Validated;
         }
         return diagnostics;
     }
 
-    documentChanged(uri: URI): void {
+    async documentChanged(uri: URI, cancelToken = CancellationToken.None): Promise<void> {
         this.langiumDocuments.invalidateDocument(uri);
         const newDocument = this.langiumDocuments.getOrCreateDocument(uri);
-        this.indexManager.update(newDocument);
-        this.langiumDocuments.all.filter(doc => this.isAffected(doc, uri)).forEach(
-            doc => {
-                this.build(doc);
-                this.indexManager.update(doc);
+        await delayNextTickAfter();
+        if (cancelToken.isCancellationRequested) {
+            return;
+        }
+        await this.buildDocument(newDocument, cancelToken);
+    }
+
+    protected async buildDocument(document: LangiumDocument, cancelToken: CancellationToken): Promise<void> {
+        const affectedDocuments = this.indexManager.getEffectedDocuments(document);
+        affectedDocuments.forEach(e => e.reset());
+        const relevantDocuments = Array.from(new Set([
+            document,
+            ...affectedDocuments,
+            // Also include all documents haven't completed the document lifecycle yet
+            ...this.langiumDocuments.all.filter(e => e.state < DocumentState.Validated)
+        ]));
+        await this.runCancelable(relevantDocuments, DocumentState.Processed, cancelToken, doc => this.process(doc));
+        await this.runCancelable(relevantDocuments, DocumentState.Indexed, cancelToken, doc => this.indexManager.update(doc, cancelToken));
+        await this.runCancelable(relevantDocuments, DocumentState.Validated, cancelToken, doc => this.validate(doc, cancelToken));
+    }
+
+    protected async runCancelable(documents: LangiumDocument[], state: DocumentState, cancelToken: CancellationToken, callback: (document: LangiumDocument) => MaybePromise<unknown>): Promise<void> {
+        if (cancelToken.isCancellationRequested) {
+            return;
+        }
+        for (const document of documents.filter(e => e.state < state)) {
+            if (cancelToken.isCancellationRequested) {
+                break;
             }
-        );
-        this.build(newDocument);
+            await callback(document);
+            await delayNextTickAfter();
+        }
     }
 
     /**
      * Process the document by running precomputations. The default implementation precomputes the scope.
      */
     protected process(document: LangiumDocument): void {
         document.precomputedScopes = this.scopeComputation.computeScope(document);
```

Comment:
I'd propose to pass the cancellation token into the scope computation.

Reply:
Hm, we hit an issue with this. If the scope computation becomes async, loading grammars becomes async as well (since they need to precompute their scope locally). This in turn makes every service that requires the grammar (parser first and foremost) async as well. I don't think that's a path we can go down. Other ideas?

Reply:
I see three solutions:

 1. Leave the scope precomputation uninterruptible.
 2. Add the grammar to the `DefaultModuleContext` so you have to resolve it _before_ you create the services.
 3. Implement a synchronous scope precomputation as a utility function specifically for the grammar language.

I'm currently favoring solution 3, that would also allow us to hard-code it to consider only grammar rules so we would get even better performance for large grammars.

Reply:
I implemented a special sync scope precomputation now like you proposed in 3. Works well 👍